### DISCUSSION TITLE
Use invite link twice.

### DISCUSSION THREAD

**Discussion by ydennisy** [supabase/supabase#6004] [General]
It is very often the case that a user will receive an invite link and click it - but maybe not complete the full sign up flow, as we require them to fill out a form (name, etc).

The implementation in supabase does not allow a user to come back and click on the same link - as it has already been verified it throws an error rather than redirecting to the necessary page.

It would be great to understand if this is by design and how the supabase team and community handle this case in their applications. I do not feel that a user should have to request another link in this case.

**Comment by burggraf** [comment]
That is most definitely by design for security purposes.  I can't think of a single company that would ever let you re-use invite links, as that can open up a huge security hole.  Typically links like this are single-use-only and also include a very limited expiration.  Otherwise if someone gets access to your email (or even copies of your old emails) they can do some pretty serious damage.

**Reply by ydennisy** [reply]
@burggraf this makes sense - and I am not asking for the invite link to be used multiple times.

What I think would be useful is for a proper error code - right now it seems there is a single error that can occur `Confirmation token not found`, would it not make sense to also have `Confirmation token already confirmed`, which would allow users of this library to pass that error onto their users in a more useful manner than just showing them a generic error message.

**Reply by burggraf** [reply]
Good point.  I'm not sure if this is also for security reasons (similar to how it's typical to not show an error message that says "invalid password" vs. "invalid credentials" because it confirms that a username exists in the database, making hacking much easier).  But it makes sense to open an issue in the GoTrue repo for this.  I'd support it.

**Reply by ydennisy** [reply]
@burggraf Ok perfect - opened an issue here https://github.com/netlify/gotrue/issues/316

**Reply by kylerummens** [reply]
For a reset-password link, would there be a security issue with having an expiration timestamp vs single-use?

I'm currently having an issue (as outlined [here](https://github.com/supabase/supabase/discussions/7791)) with enterprise email systems that will parse emails and open all the links. This obviously invalidates the reset-password link so by the time the email actually hits the user's inbox, it's not usable.

**Reply by naripok** [reply]
We are still facing issues with the invalidation behaviour here. As described by @kylerummens, we have enterprise users with email systems that will open links, invalidating any invitation/confirmation links we send them.

A work-around being considered is to use a proxy page in which the user will land after clicking the email's link, which would contain a button with the actual link and a captcha so to prevent the automated systems to invalidate the link. While it seems possible to do this by changing the email template in supabase studio UI, it is not possible to directly change it in the development studio version from supabase-cli, making testing very hard.

I've found the templates [here](https://github.com/supabase/supabase/blob/8f217deeb6b4d56f76fc848ae240e25448518eca/studio/pages/api/auth/%5Bref%5D/config.ts#L113-L126), but they are hard-coded. So, @burggraf, do you know of any way to change/configure the email templates on the dev server without having to change the flow and use docker and manually change the hard-coded value instead of supabase-cli?

Thanks and sorry for bothering! :)

**Reply by ydennisy** [reply]
@burggraf We really need your help on this point. Would appreciate you taking a look.

**Reply by grimuz** [reply]
@burggraf this problem is also relevant in our company using supabase. We also have the following problem. Quote:

> we have enterprise users with email systems that will open links, invalidating any invitation/confirmation links we send them.

**Comment by OverGreed** [comment]
Probably it's might be a good idea to allow them to use this link until he set a password.
Because after using this link use will get only a one-time session and that's all.

**Reply by n-hallberg** [reply]
Yes, that would be great! I would also love to see this behavior. Currently, I'm considering developing these functionalities via my own mail service, but a few more options in the builtin server would still be cool though. :)
